0304~0308 육다빈 (0304 치즈 완료, 0308 3문제)

Dabisix opened the pull request:
# 0304 치즈
- DFS로 풀이
- 1차 시도 : 제출 시 스택 오버플로 발생으로 실패
    - DFS부분에서 재귀를 빠져나오지 못하고 무한 루프를 도는 듯 했다
    - 내부 체크 시, 처음에 탐색하는 중심부에서 다음으로 넘어갈 때, 넘어온 칸에서는 이전 칸을 제외하고 나머지 칸에 대해 탐색하는 식으로 작성해서, 칸끼리 계속 돌고 돌았을 것으로 추정
    - 알고리즘 로직에 문제가 있는 듯 하여 다른 방법을 찾았다
- 2차시도 : 매 회차마다 외부를 탐색하는 방식으로 변경
    - 생각해보니 외곽은 항상 외부라고 정해져 있어 시작점을 잡기도 좋고 dfs한번만에 모든 외부를 탐색할 수 있어, 내부보다 외부 탐색이 더욱 적절했음을 알게됨
    - 매 회차마다 외부를 표시하는 기호를 바꿔가며 외부 마킹
    
  <aside>
  🌟 고려 사항
  
  - 외부 마킹을 char로 하다 보니, 최대 마킹 횟수에 제한이 있었음
  - char++ 해주는 형식으로 했는데, 최대 약 6만까지 있었으나, dfs를 그정도로 돌리기 전에 문제가 풀린다고 생각하여 그냥 씀
  </aside>

Dabisix committed:
[백준] 2638 치즈 (완)

- DFS로 풀이
- 1차 시도 때, 제출 시 스택 오버플로 발생으로 실패
    - DFS부분에서 재귀를 빠져나오지 못하고 무한 루프를 도는 듯 했다
    - 내부 체크 시, 처음에 탐색하는 중심부에서 다음으로 넘어갈 때, 넘어온 칸에서는 이전 칸을 제외하고 나머지 칸에 대해 탐색하는 식으로 작성해서, 칸끼리 계속 돌고 돌았을 것으로 추정
    - 알고리즘 로직에 문제가 있는 듯 하여 다른 방법을 찾았다
- 2차시도 : 매 회차마다 외부를 탐색하는 방식으로 변경
    - 생각해보니 외곽은 항상 외부라고 정해져 있어 시작점을 잡기도 좋고 dfs한번만에 모든 외부를 탐색할 수 있어, 내부보다 외부 탐색이 더욱 적절했음을 알게됨
    - 매 회차마다 외부를 표시하는 기호를 바꿔가며 외부 마킹
    
    <aside>
    🌟 고려 사항
    
    - 외부 마킹을 char로 하다 보니, 최대 마킹 횟수에 제한이 있었음
    - char++ 해주는 형식으로 했는데, 최대 약 6만까지 있었으나, dfs를 그정도로 돌리기 전에 문제가 풀린다고 생각하여 그냥 씀
    </aside>

Dabisix committed:
[백준] 20364 부동산 다툼

- 이진트리에다가, 노드의 숫자가 배정되는 룰이 있기에 따로 노드 값을 저장하지 않았음
- 목적지에서부터 출발지까지 역으로 탐색하여 막히는 부분 체크
- 이미 점유된 땅을 점유하는 경우를 체크하지 못해 한번 틀림

Dabisix commented:
# 0308 부동산 다툼
- 이진트리에다가, 노드의 숫자가 배정되는 룰이 있기에 따로 노드 값을 저장하지 않았다
- 목적지에서부터 출발지까지 역으로 탐색하여 막히는 부분 체크
- 이미 점유된 땅을 점유하는 경우를 체크하지 못해 한번 틀림

Dabisix commented:
# 0308 봄버맨
- 단순 반복문 사용
- 매 초마다 전체 배열을 돌면서 하나씩 상태변화 시킴
- 폭탄의 경우 예외적으로, 전체 배열 상태변화 후 한번에 터트림

Dabisix committed:
[백준] 16918 봄버맨

단순 반복문 사용

Dabisix commented:
# 0308 농장 관리
- bfs로 팔방탐색하여 풀이
- 잊어가던 bfs 다시 단단하게 할 수 있는 좋은 시간이었습니다

Dabisix committed:
[백준] 1245 농장 관리

bfs로 팔방탐색

hanyoonseok commented on 육다빈/0308/b1245.java:
```diff
@@ -0,0 +1,86 @@
+				if(map[i][j] == peek || map[i][j] == 0) continue;	// 이미 탐색한 산봉우리나, 평지 제외
+				visit = new boolean[N][M];
+				List<Node> list = bfs(i, j);	// 탐색된 산봉우리 리스트
+				if(!list.isEmpty()) {
```
코드 깔끔하고 읽기도 좋네ㅋㅋㅋ 근데 이 부분 로직을 bfs 함수 안에서 처리하면 main 메서드 안에서 List를 한번 더 선언해주지 않아도 되고 반환형도 void로 쓸 수 있지 않아?!

Dabisix replied:
오 그러네 ?? 그게 더 좋겠다 피드백 감삼당

hanyoonseok commented on 육다빈/0308/b16918.java:
```diff
@@ -0,0 +1,55 @@
+			for(int i=0; i<R; i++) {
+				for(int j=0; j<C; j++) {
+					switch(map[i][j]) {
+					case '.': map[i][j] = '-';  break;	// 초기 시작 시 1초 쉬기
```
switch문이 가독성이 떨어진다고 생각하고 있었는데, 이렇게 구분하니까 훨씬 깔끔한듯?? 라인수도 짧고 쓸데없는게 많이 정제된 느낌

Dabisix replied:
마자 매번 break 써줘야 해서 잘 안쓰긴 하는데, 이번처럼 값만 바꿔줄 때 자주 쓰기 좋더라구

Dabisix committed:
[백준] 1245 농장 관리 (피드백 반영)

변경 사항 : 윤석님의 피드백 반영
- bfs로 list를 반환하여 main에서 실행하던 코드를, bfs 내부로 넣음
개선 사항
- 메모리 낭비 감소

Dabisix commented:
# 농장 관리 - 윤석님의 피드백 반영
- 변경 사항 : bfs로 list를 반환하여 main에서 실행하던 코드를, bfs 내부로 넣음
- 개선된 점 : 메모리 낭비 감소

yoochangseon commented:
### 봄버맨: 제일 코드 깔끔하게 잘 짠듯!
### 부동산 관리: 패스ㅋㅋㅋ
### 농장관리 : BFS로 푼 것과 난 DFS 차이 정도..?
